src/stirling/source_connectors/perf_profiler/pprof/pprof.cc

```diff
@@ -0,0 +1,148 @@
+
+#include <gflags/gflags.h>
+
+#include <absl/strings/str_split.h>
+#include <vector>
+
+DECLARE_uint32(stirling_profiler_stack_trace_sample_period_ms);
+
+namespace px {
+namespace stirling {
+
+PProfProfile CreatePProfProfile(const uint32_t num_cpus, const histo_t& histo) {
+  // Info on the pprof proto format:
+  // https://github.com/google/pprof/blob/main/proto/profile.proto
+
+  // period_ns will be used when populating the nanos count.
+  const uint64_t period_ms = FLAGS_stirling_profiler_stack_trace_sample_period_ms;
+  const uint64_t period_ns = period_ms * 1000 * 1000;
+
+  // Tracks which strings have been inserted into the profile.
+  absl::flat_hash_map<std::string, uint64_t> strings;
+
+  // This is the pprof profile.
+  ::perftools::profiles::Profile profile;
+
+  // The pprof profiles start by describing their sample types. For CPU profiling,
+  // the convention is to describe two different metrics:
+  // "samples" with units of "count", and
+  // "cpu" with units of "nanoseconds".
+  // Note, the first entry in the strings table is required to be an empty string.
+  auto sample_type = profile.add_sample_type();
+  sample_type->set_type(1);
+  sample_type->set_unit(2);
+  profile.add_string_table("");
+  profile.add_string_table("samples");
+  profile.add_string_table("count");
+  sample_type = profile.add_sample_type();
+  sample_type->set_type(3);
+  sample_type->set_unit(4);
+  profile.add_string_table("cpu");
+  profile.add_string_table("nanoseconds");
+
+  // State variables useful as we build the pprof profile message.
+  // No locations messages exist (yet); next_location_id starts at 1.
+  // We already added some strings; next_string_id starts at 5.
+  uint64_t next_location_id = 1;
+  uint64_t next_string_id = 5;
+
+  // To build the profile, we iterate over the stack traces  histogram.
+  for (const auto& [stack_trace_str, count] : histo) {
+    // Each histo entry will be recorded as a new sample.
+    auto sample = profile.add_sample();
+
+    // That sample will record its count and time in nanos.
+    const uint64_t nanos = count * period_ns / num_cpus;
+    sample->add_value(count);
+    sample->add_value(nanos);
+
+    // Our stack traces are symbolized like so: main;foo;bar
+    // thus, we split on ';' to iterate over the individual symbols.
+    const std::vector<std::string_view> symbols = absl::StrSplit(stack_trace_str, ";");
+
+    // Iterate over the symbols in this stack trace.
+    // Each symbol will be added to the sample as a "location" message, which in turn refers to a
+    // string in the strings table (tracked in map "strings") (more details below).
+    // Note, because of how we built the stack trace string and how the pprof profile is
+    // organized, we iterate in reverse order.
+    for (auto symbols_iter = symbols.rbegin(); symbols_iter != symbols.rend(); ++symbols_iter) {
+      // For convenience.
+      const auto& symbol = *symbols_iter;
```

Comment:
Feel free to drop. Given the depth of the other comments and usefulness, I think this one could be removed.